library/hls/src/main/java/com/google/android/exoplayer2/source/hls/playlist/DefaultHlsPlaylistTracker.java

```diff
@@ -550,6 +559,18 @@ public boolean isSnapshotValid() {
           || lastSnapshotLoadMs + snapshotValidityDurationMs > currentTimeMs;
     }
 
+    public boolean isSnapshotFresh() {
+      if (playlistSnapshot == null) {
+        return false;
+      }
+      long currentTimeMs = SystemClock.elapsedRealtime();
+      long updateDelta = currentTimeMs - lastSnapshotLoadMs;
+      return playlistSnapshot.hasEndTag
+          || playlistSnapshot.playlistType == HlsMediaPlaylist.PLAYLIST_TYPE_EVENT
+          || playlistSnapshot.playlistType == HlsMediaPlaylist.PLAYLIST_TYPE_VOD
+          || updateDelta <= C.usToMs(playlistSnapshot.targetDurationUs * 2);
+    }
```

Comment:
Possible the `TYPE_EVENT` should also be checked for freshness rather then assuming it is.

Note once it has updated once it may be overkill to check delta < target duration (as this is already covered by the stuck playlist check for primary), so just is primary and has updated (`lastSnapshotLoadMs != 0` or just `playlistSnapshot != null`).  If so, this method may not be needed.